run_diff.go

```diff
@@ -0,0 +1,142 @@
+// limitations under the License.
+
+package wptdashboard
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/datastore"
+	"google.golang.org/appengine/urlfetch"
+)
+
+type platformAtRevision struct {
+	Platform string
+	Revision string
+}
+
+func parsePlatformAtRevisionSpec(spec string) (platformAtRevision platformAtRevision, err error) {
+	pieces := strings.Split(spec, "@")
+	if len(pieces) > 2 {
+		return platformAtRevision, errors.New("invalid platform@revision spec: " + spec)
+	}
+	platformAtRevision.Platform = pieces[0]
+	if len(pieces) < 2 {
+		// No @ is assumed to be the platform only.
+		platformAtRevision.Revision = "latest"
+	} else {
+		platformAtRevision.Revision = pieces[1]
+	}
+	if IsBrowserName(platformAtRevision.Platform) {
+		return platformAtRevision, nil
+	}
+	return platformAtRevision, errors.New("Platform " + platformAtRevision.Platform + " not found")
+}
+
+func fetchRunResultsJSONForParam(
+	ctx context.Context, r *http.Request, revision string) (results map[string][]int, err error) {
+	var spec platformAtRevision
+	if spec, err = parsePlatformAtRevisionSpec(revision); err != nil {
+		return nil, err
+	}
+	return fetchRunResultsJSONForSpec(ctx, r, spec)
+}
+
+func fetchRunResultsJSONForSpec(
+	ctx context.Context, r *http.Request, revision platformAtRevision) (results map[string][]int, err error) {
+	var run TestRun
+	if run, err = fetchRunForSpec(ctx, revision); err != nil {
+		return nil, err
+	} else if (run == TestRun{}) {
+		return nil, nil
+	}
+	return fetchRunResultsJSON(ctx, r, run)
+}
+
+func fetchRunForSpec(ctx context.Context, revision platformAtRevision) (TestRun, error) {
+	baseQuery := datastore.
+		NewQuery("TestRun").
+		Order("-CreatedAt").
+		Limit(1)
+
+	var results []TestRun
+	query := baseQuery.
+		Filter("BrowserName =", revision.Platform)
+	if revision.Revision != "latest" {
+		query = query.Filter("Revision = ", revision.Revision)
+	}
+	if _, err := query.GetAll(ctx, &results); err != nil {
+		return TestRun{}, err
+	}
+	if len(results) < 1 {
+		return TestRun{}, nil
+	}
+	return results[0], nil
+}
+
+func fetchRunResultsJSON(ctx context.Context, r *http.Request, run TestRun) (results map[string][]int, err error) {
+	client := urlfetch.Client(ctx)
+	url := run.ResultsURL
+	if strings.Index(url, "/") == 0 {
+		reqURL := *r.URL
+		reqURL.Path = url
+	}
+	var resp *http.Response
+	if resp, err = client.Get(url); err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("%s returned HTTP status %d", url, resp.StatusCode)
+	}
+	var body []byte
+	if body, err = ioutil.ReadAll(resp.Body); err != nil {
+		return nil, err
+	}
+	if err = json.Unmarshal(body, &results); err != nil {
+		return nil, err
+	}
+	return results, nil
+}
+
+func diffResults(before map[string][]int, after map[string][]int) map[string][]int {
```

Comment:
@mdittmer, here are a bunch of loops over test results, this is the shape of things I was imagining would be involved in computing metrics. Just FYI, let's keep going in the design doc.